### PR TITLE
Post patch methods

### PR DESCRIPTION
**Pull request by LadaRudnitckaia**
#21 POST

### CODE REVIEW

File: R/kobo-client.R [hunk 1]
```diff
@@ -35,10 +36,13 @@ KoboClient <- R6::R6Class("KoboClient",
       super$initialize(
         url = base_url,
         headers = list(
-          Authorization = paste0("Token ", kobo_token)
+          Authorization = paste0("Token ", kobo_token),
+          Accept = "application/json"
+          #"content-type" = "application/json"
```
Comment (MKyhos) [inline R/kobo-client.R]: In case the commented header element (`content-type`)  is not needed, this line might be deleted.

```suggestion
```

File: R/kobo-client.R [hunk 2]
```diff
@@ -54,12 +58,45 @@ KoboClient <- R6::R6Class("KoboClient",
...
+    #' @param ... crul-options. Additional option arguments, see
+    #'  [`crul::HttpClient`] for reference
+    #' @return Returns a list, parsed from the HttpResponse JSON object.
+    post = function(path, body, ...) {
+      # Check the body's format
+      # if (is.list(body)) {
+      #     body <- list_as_json_char(body)
+      # } else if (!is.character(body)) {
+      #     stop("The body should be either an R list (will be automatically
+      #          converted to JSON-like string) or a JSON-like string.")}
```
Comment (MKyhos) [inline R/kobo-client.R]: If checks for the body format are indeed obsolete here, these lines might be deleted, or is this to be figured out? :)

File: R/kobo-client.R [hunk 3]
```diff
@@ -54,12 +58,45 @@ KoboClient <- R6::R6Class("KoboClient",
...
+      res <- super$post(path = path, body = body, ...)
+
       # Perform additional checks, json to list parsing
       res$raise_for_status()
-      res$raise_for_ct_json()
-      res$parse("UTF-8") %>%
-        jsonlite::fromJSON()
+      # res$raise_for_ct_json()
+      # res$parse("UTF-8") %>%
+      #     jsonlite::fromJSON()
```
Comment (MKyhos) [inline R/kobo-client.R]: Same goes for parsing and `raise_for_ct_json()`. 
I remember this to be a matter of discussion, I assume it is settled now that the full response object shall be returned from `KoboClient$post()`, not a parsed JSON`? @ all :slightly_smiling_face:

File: R/kobo-client.R [hunk 4]
```diff
@@ -54,12 +58,45 @@ KoboClient <- R6::R6Class("KoboClient",
...
+      return(res)
     }
-  )
+  ) # <end public>
```
Comment (MKyhos) [inline R/kobo-client.R]: I added those tags in order to make it make it more readable, in case of long `private` or `public` lists - I hope this is okay!

File: R/kobo.R [hunk 5]
```diff
@@ -63,21 +44,115 @@ Kobo <- R6::R6Class("Kobo",
...
+    #' @description
+    #' High-level POST request to create an empty asset. `assets/` endpoint
+    #' (due to default to `v2`, no further specification is needed).
+    #' @param name character. Name of the new asset.
+    #' @param description character. Optional.
+    #' @param sector character. Optional.
+    #' @param country character. Optional.
+    #' @param share_metadata boolean. Optional.
+    #' @param asset_type character. Type of the new asset. Can be
+    #' "block", "question", "survey", "template".
```
Comment (MKyhos) [inline R/kobo.R]: Maybe you could copy / adapt the `@return` tag from the `$post()` method here, that informs about the response returning behavior of the `$create_asset()` method

File: R/kobo.R [hunk 6]
```diff
@@ -63,21 +44,115 @@ Kobo <- R6::R6Class("Kobo",
...
+      res_list <- self$get("assets/")
+    },
+
+    #' @description
+    #' High-level POST request to clone an asset. `assets` endpoint
+    #' (due to default to `v2`, no further specification is needed).
+    #' @param clone_from character. UID of the asset to be cloned.
+    #' @param name character. Name of the new asset.
+    #' @param asset_type character. Type of the new asset. Can be
+    #' "block", "question", "survey", "template".
```
Comment (MKyhos) [inline R/kobo.R]: Same here about the `@return` 
Maybe: rename `name` to `new_name`? 

And has the `asset_type` to be the same as the type of the cloned asset? Maybe we could catch this somehow from the API directly? :thinking:

File: R/kobo.R [hunk 7]
```diff
@@ -63,21 +44,115 @@ Kobo <- R6::R6Class("Kobo",
...
+        "name" = name,
+        "asset_type" = asset_type)
+      self$post("assets/", body = body)
+    },
+
+    #' @description
+    #' High-level POST request to deploy an asset.
+    #' `assets/{uid}/deployment/` endpoint (due to
+    #' default to `v2`, no further specification is needed).
+    #' @param uid character. UID of the asset to be deployed.
```
Comment (MKyhos) [inline R/kobo.R]: `@return`

File: R/list_as_json_char.R [hunk 8]
```diff
@@ -0,0 +1,24 @@
+#' Helper function to convert R list to JSON-like string
```
Comment (MKyhos) [inline R/list_as_json_char.R]: For this function, you could consider moving it into the `utils.R` file, instead of it having its own source file, or do you prefer it this way? :)

File: tests/testthat/test-kobo.R [hunk 9]
```diff
@@ -100,3 +100,82 @@ test_that("non-existing kobo host throws error", {
   kobo <- Kobo$new(base_url_v2 = "https://nokobohere.correlaid.org", kobo_token = Sys.getenv("KBTBR_TOKEN"))
   expect_error(kobo$get("assets/"), regexp = "^SSL.+certificate.+")
 })
+
```
Comment (MKyhos) [inline tests/testthat/test-kobo.R:125]: Nice work on the tests!! 

(just a very minor comment RE code styling for both test files, breaking longer lines into multiple lines such that it stays roughly within the 80 characters +/- could be conducted, but thus is not related to the functionality of course :see_no_evil: )
See the styler comment also below

File: tests/testthat/test-kobo.R [hunk 10]
```diff
@@ -100,3 +100,82 @@ test_that("non-existing kobo host throws error", {
...
+    vcr::use_cassette("kobo-clone-assets-simple-post", {
+        kobo <- Kobo$new(base_url_v2 = BASE_URL, kobo_token = Sys.getenv("KBTBR_TOKEN"))
+        clone_asset <- kobo$post("assets/", body = list("clone_from" = "a84jmwwdPEsZMhK7s2i4SL",
+                                                        "name" = "vcr_test_name",
+                                                        "asset_type" = "survey"))
+    })
+    expect_equal(clone_asset$url,"https://kobo.correlaid.org/api/v2/assets/")
+    expect_equal(clone_asset$method,"post")
+    expect_equal(clone_asset$status_code,201)
+    expect_equal(clone_asset$success(),TRUE)
```
Comment (MKyhos) [inline tests/testthat/test-kobo.R]: Here also the `testthat::expect_true()`  function could be used :)

File: tests/testthat/test-kobo.R [hunk 11]
```diff
@@ -100,3 +100,82 @@ test_that("non-existing kobo host throws error", {
...
+    expect_equal(clone_asset$status_http()$message,"Created")
+    expect_equal(clone_asset$status_http()$explanation,"Document created, URL follows")
+})
+
+test_that("kobo$clone_asset can clone assets", {
+    vcr::use_cassette("kobo-post-clone-asset", {
+        kobo <- suppressWarnings(Kobo$new(base_url_v2 = BASE_URL, kobo_token = Sys.getenv("KBTBR_TOKEN")))
```
Comment (MKyhos) [inline tests/testthat/test-kobo.R]: Which purpose does the `suppressWarnings()` have here? :)

File: R/kobo.R [hunk 12]
```diff
@@ -63,21 +44,115 @@ Kobo <- R6::R6Class("Kobo",
...
+    #' (due to default to `v2`, no further specification is needed).
+    #' @param name character. Name of the new asset.
+    #' @param description character. Optional.
+    #' @param sector character. Optional.
+    #' @param country character. Optional.
+    #' @param share_metadata boolean. Optional.
+    #' @param asset_type character. Type of the new asset. Can be
+    #' "block", "question", "survey", "template".
+    create_asset = function(name,description,sector,
+                            country,share_metadata,asset_type) {
```
Comment (MKyhos) [inline R/kobo.R]: Default values for sector / country, share_metadata could be considered, to make them optional / empty by default or so. Otherwise users have to set them explicitly every time they call the  method.
(once everything works)

File: R/kobo.R [hunk 13]
```diff
@@ -63,21 +44,115 @@ Kobo <- R6::R6Class("Kobo",
...
+        private$session_v1$post(path = paste0("api/v1/", path), body = body)
+      } else {
+        usethis::ui_stop(
+          "Invalid version. Must be either v1 or v2.
+          Come back in a couple of years."
+        )
+      }
+    },
+
     #' @description
```
Comment (MKyhos) [inline R/kobo.R]: An idea I had regarding the asset list / data:
Should we store them in a field of the `Kobo` object such that we can access it also from other methods? tbd

File: R/kobo-client.R [hunk 14]
```diff
@@ -54,12 +58,45 @@ KoboClient <- R6::R6Class("KoboClient",
...
+    #' the HttpResponse object on status, that it is of type
+    #' `application/json`, checks whether the body is a json-like string,
+    #' and if not, converts it to the json-like string, and parses the
+    #' response text subsequently from JSON to R list representation.
+    #' @param path character. Path component of the endpoint.
+    #' @param body R list or json-like string. A data payload to be sent
+    #' to the server.
+    #' @param ... crul-options. Additional option arguments, see
+    #'  [`crul::HttpClient`] for reference
+    #' @return Returns a list, parsed from the HttpResponse JSON object.
```
Comment (MKyhos) [inline R/kobo-client.R]: Update docs, as currently the raw `response` is returned, and not a parsed list :notebook: 
```suggestion
    #' @return Returns an object of class `crul::HttpResponse`.
```

File: tests/testthat/test-kobo.R [hunk 15]
```diff
@@ -100,3 +100,82 @@ test_that("non-existing kobo host throws error", {
...
+
+test_that("kobo$deploy_asset can deploy assets", {
+    vcr::use_cassette("kobo-post-deploy-asset", {
+        kobo <- suppressWarnings(Kobo$new(base_url_v2 = BASE_URL, kobo_token = Sys.getenv("KBTBR_TOKEN")))
+        deploy_asset <- kobo$deploy_asset(uid = "aQVGH8G68EP737tDBABRwC")
+    })
+    expect_equal(deploy_asset$url,
+                 "https://kobo.correlaid.org/api/v2/assets/aQVGH8G68EP737tDBABRwC/deployment/")
+    expect_equal(deploy_asset$method,"post")
+    expect_equal(deploy_asset$status_code,200) # i don't understand why not 201. but with 200 it successfully deployed
```
Comment (MKyhos) [inline tests/testthat/test-kobo.R]: Hah super weird! :mag: 
Maybe this is something we could bring to the kpi github repo? :thinking:

File: tests/testthat/test-kobo-client.R [hunk 16]
```diff
@@ -28,8 +28,24 @@ test_that("Kobo Client can fetch assets", {
...
+    vcr::use_cassette("kobo-client-clone-assets-simple-post", {
+        koboclient <- KoboClient$new(base_url = BASE_URL, kobo_token = Sys.getenv("KBTBR_TOKEN"))
+        clone_asset <- koboclient$post("api/v2/assets/",
+                                       body = list("clone_from" = "a84jmwwdPEsZMhK7s2i4SL",
+                                                   "name" = "vcr_test_name", "asset_type" = "survey"))
+    })
+    expect_equal(clone_asset$url,"https://kobo.correlaid.org/api/v2/assets/")
+    expect_equal(clone_asset$method,"post")
+    expect_equal(clone_asset$status_code,201)
+    expect_equal(clone_asset$success(),TRUE)
```
Comment (MKyhos) [inline tests/testthat/test-kobo-client.R]: There is a `testthat::expect_true()` function which might come handy here :) 
```suggestion
    expect_truel(clone_asset$success())
```

**Review by MKyhos — CHANGES_REQUESTED** [review]
Hey, cool & amazing work! :) 
In a first attempt, I added several comments we could discuss / resolve :pray:

**Comment by MKyhos** [comment]
Also a general remark / question regarding the `Kobo` methods 
- `clone_asset()`
- `deploy_asset()`
- `create_asset()`
- `import_xlsx_form()`

Currently, they all return a `crul::HttpResponse` object (and raise an error based on the status). I have only little to no experience with working with the code, and from my naive viewpoint it would make sense for the enduser to just receive the full response object for debugging purposes, and otherwise just receive a printed statement? 🤔 
Like:
```
CorrelKobo <- Kobo$new(...)

CorrelKobo$clone_asset(correct arguments, debug = FALSE)
#> Print success message, return nothing

CorrelKobo$clone_asset(wrong arguments / or wrong server config, debug = FALSE) 
#> Print error message, return nothing  

CorrelKobo$clone_asset(any arguments, debug = TRUE)
#> Print error message, return response object
```

**Comment by MKyhos** [comment]
And depending on the teams preferences on indention,
```
style_pkg(style = tidyverse_style, indent_by = 4) 
``` 
(requires the `styler` package)

File: R/kobo.R [hunk 17]
```diff
@@ -63,21 +44,115 @@ Kobo <- R6::R6Class("Kobo",
...
+    #' (due to default to `v2`, no further specification is needed).
+    #' @param name character. Name of the new asset.
+    #' @param file character. The path to the file containing the XLS form.
+    import_xls_form = function(name,file) {
+      body = list(
+        "name"=name,
+        "library"="false",
+        "file"=crul::upload(file))
+      self$post("imports/", body = body)
     }
```
Comment (MKyhos) [inline R/kobo.R]: Maybe `file_path` is a clearer argument name?

```suggestion
    #' @param file_path  character. The path to the XLS form file.
    import_xls_form = function(name, file_path) {
      body = list(
        "name"=name,
        "library"="false",
        "file"=crul::upload(file_path))
      self$post("imports/", body = body)
    }
```

File: R/kobo-client.R (hunk 2 above)
Comment (LadaRudnitckaia) [reply R/kobo-client.R]: done

File: R/kobo-client.R (hunk 3 above)
Comment (LadaRudnitckaia) [reply R/kobo-client.R]: true. done

File: R/kobo-client.R (hunk 4 above)
Comment (LadaRudnitckaia) [reply R/kobo-client.R]: sure

File: R/kobo.R (hunk 5 above)
Comment (LadaRudnitckaia) [reply R/kobo.R]: added to all functions in kobo

File: R/list_as_json_char.R (hunk 8 above)
Comment (LadaRudnitckaia) [reply R/list_as_json_char.R]: I think it can be deleted. seems that we don't need to convert r lists to son like strings

Comment (LadaRudnitckaia) [reply R/list_as_json_char.R]: I removed it

File: tests/testthat/test-kobo.R (hunk 9 above)
Comment (LadaRudnitckaia) [reply tests/testthat/test-kobo.R:125]: I broke some of them, let me know if they are still too long

File: tests/testthat/test-kobo.R (hunk 10 above)
Comment (LadaRudnitckaia) [reply tests/testthat/test-kobo.R]: done

File: tests/testthat/test-kobo.R (hunk 11 above)
Comment (LadaRudnitckaia) [reply tests/testthat/test-kobo.R]: it actually dines't help. I wanted to avoid the messages about api version during tests but they were still returned

File: R/kobo.R (hunk 12 above)
Comment (LadaRudnitckaia) [reply R/kobo.R]: maybe I can comment this method for now?.. for the first merge

File: R/kobo-client.R (hunk 14 above)
Comment (LadaRudnitckaia) [reply R/kobo-client.R]: done

File: tests/testthat/test-kobo.R (hunk 15 above)
Comment (LadaRudnitckaia) [reply tests/testthat/test-kobo.R]: maybe

**Comment by LadaRudnitckaia** [comment]
as for the response - I'm not sure. This crul::HttpResponse is not very big actually. why not to keep it?

File: R/kobo.R (hunk 6 above)
Comment (LadaRudnitckaia) [reply R/kobo.R]: changed to new_name. I think we would need to do a get request inside of the post to get this asset type

File: R/kobo.R (hunk 13 above)
Comment (LadaRudnitckaia) [reply R/kobo.R]: I don't get it. Could you please elaborate? :)

File: R/kobo.R (hunk 12 above)
Comment (MKyhos) [reply R/kobo.R]: Yes, I agree! Commenting it out or even exclude it in total from the current PR, and instead write our current findings into a separate issue. Whatever you prefer :)

File: R/list_as_json_char.R (hunk 8 above)
Comment (MKyhos) [reply R/list_as_json_char.R]: Yes, agreed. It might depend on our future success (hopefully, there is any) on the asset creation, where we might have to provide the `settings` value as a json string... But I agree, we can add potentially needed helper functions in the future.

File: tests/testthat/test-kobo.R (hunk 10 above)
Comment (MKyhos) [reply tests/testthat/test-kobo.R]: Thank you :)

File: tests/testthat/test-kobo.R (hunk 11 above)
Comment (MKyhos) [reply tests/testthat/test-kobo.R]: Okay! In that case you could use `suppressMessages()`  for that purpose! In general, I would argue to avoid suppressing warnings inside tests, as we could miss some other warnings (unintended).

File: R/kobo.R (hunk 13 above)
Comment (MKyhos) [reply R/kobo.R]: I was just wondering whether we should store the asset list as a field of the Kobo object, that is, first declare a (private, or public) filed `asset_list`, and then inside this method include something like ` self$asset_list <- asset_list`  (or the same with private). 

Maybe this is something we could discuss? Just out of my gut, I thought this could be something useful. For example, linked to my comment on the clone asset method ("could we infer the asset type for the new asset from the type of the asset to be cloned?") could make use of such an internally stored asset list. 

But there are probably several other considerations, e.g. how often to update the internal asset list, etc. 

**TL;DR** I will try to elaborate my thoughts further until the next meeting, and we can ignore it for this PR.

File: R/kobo-client.R (hunk 14 above)
Comment (MKyhos) [reply R/kobo-client.R]: Thank you )

File: R/kobo.R (hunk 6 above)
Comment (MKyhos) [reply R/kobo.R]: Yes you are right! See my comment on storing `asset_list` also in the object as a field, maybe we can discuss this in the future. But :heavy_check_mark:  for now :)

File: R/kobo.R (hunk 5 above)
Comment (MKyhos) [reply R/kobo.R]: thanks )

File: tests/testthat/test-kobo.R (hunk 9 above)
Comment (MKyhos) [reply tests/testthat/test-kobo.R:125]: Looks good to me! :)

**Review by MKyhos — APPROVED** [review]
